src/m3ninx/index/segment/builder/multi_segments_multi_key_postings_list_iter.go

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package builder
+
+import (
+	"bytes"
+
+	"github.com/m3db/m3/src/m3ninx/index"
+	"github.com/m3db/m3/src/m3ninx/index/segment"
+	"github.com/m3db/m3/src/m3ninx/postings"
+	"github.com/m3db/m3/src/m3ninx/postings/roaring"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	bitmap "github.com/m3db/pilosa/roaring"
+)
+
+var _ segment.FieldsPostingsListIterator = &multiKeyPostingsListIterator{}
+
+type multiKeyPostingsListIterator struct {
+	err                    error
+	firstNext              bool
+	closeIters             []keyIterator
+	iters                  []keyIterator
+	currIters              []keyIterator
+	currReaders            []index.Reader
+	segments               []segment.Segment
+	currFieldPostingsList  postings.MutableList
+	currFieldPostingsLists []postings.List
+}
+
+func newMultiKeyPostingsListIterator() *multiKeyPostingsListIterator {
+	b := bitmap.NewBitmapWithDefaultPooling(defaultBitmapContainerPooling)
+	i := &multiKeyPostingsListIterator{
+		currFieldPostingsList: roaring.NewPostingsListFromBitmap(b),
+	}
+	i.reset()
+	return i
+}
+
+func (i *multiKeyPostingsListIterator) reset() {
+	i.firstNext = true
+	i.currFieldPostingsList.Reset()
+
+	for j := range i.closeIters {
+		i.closeIters[j] = nil
+	}
+	i.closeIters = i.closeIters[:0]
+
+	for j := range i.iters {
+		i.iters[j] = nil
+	}
+	i.iters = i.iters[:0]
+
+	for j := range i.currIters {
+		i.currIters[j] = nil
+	}
+	i.currIters = i.currIters[:0]
```

Comment:
Should also clear out all of `i.currReaders` too yeah?

Reply:
I'm going to just clear and reset like u mentioned below inside of `Next()`?

Reply:
Sounds good.